crates/bevy_asset/src/reflect.rs

```diff
@@ -0,0 +1,263 @@
+    }
+
+    /// Equivalent of [`Assets::len`]
+    #[allow(clippy::len_without_is_empty)] // clippy expects the `is_empty` method to have the signature `(&self) -> bool`
+    pub fn len(&self, world: &World) -> usize {
+        (self.len)(world)
+    }
+
+    /// Equivalent of [`Assets::is_empty`]
+    pub fn is_empty(&self, world: &World) -> bool {
+        self.len(world) == 0
+    }
+
+    /// Equivalent of [`Assets::ids`]
+    pub fn ids<'w>(&self, world: &'w World) -> impl Iterator<Item = HandleId> + 'w {
+        (self.ids)(world)
+    }
+}
+
+impl<A: Asset + FromReflect> FromType<A> for ReflectAsset {
+    fn from_type() -> Self {
+        ReflectAsset {
+            type_uuid: A::TYPE_UUID,
+            assets_resource_type_id: TypeId::of::<Assets<A>>(),
+            get: |world, handle| {
+                let assets = world.resource::<Assets<A>>();
+                let asset = assets.get(&handle.typed());
+                asset.map(|asset| asset as &dyn Reflect)
+            },
+            get_mut: |world, handle| {
+                let assets = world.resource_mut::<Assets<A>>().into_inner();
+                let asset = assets.get_mut(&handle.typed());
+                asset.map(|asset| asset as &mut dyn Reflect)
+            },
+            get_unchecked_mut: |world, handle| {
+                let assets = unsafe {
+                    world
+                        .get_resource_unchecked_mut::<Assets<A>>()
+                        .unwrap()
+                        .into_inner()
+                };
+                let asset = assets.get_mut(&handle.typed());
+                asset.map(|asset| asset as &mut dyn Reflect)
+            },
+            add: |world, value| {
+                let mut assets = world.resource_mut::<Assets<A>>();
+                let value: A = FromReflect::from_reflect(value)
+                    .expect("could not call `FromReflect::from_reflect` in `ReflectAsset::insert`");
+                assets.add(value).into()
+            },
+            set: |world, handle, value| {
+                let mut assets = world.resource_mut::<Assets<A>>();
+                let value: A = FromReflect::from_reflect(value)
+                    .expect("could not call `FromReflect::from_reflect` in `ReflectAsset::insert`");
+                assets.set(handle, value).into()
+            },
+            len: |world| {
+                let assets = world.resource::<Assets<A>>();
+                assets.len()
+            },
+            ids: |world| {
+                let assets = world.resource::<Assets<A>>();
+                Box::new(assets.ids())
+            },
+            remove: |world, handle| {
+                let mut assets = world.resource_mut::<Assets<A>>();
+                let value = assets.remove(handle);
+                value.map(|value| Box::new(value) as Box<dyn Reflect>)
+            },
+        }
+    }
+}
+
+/// Reflect type data struct relating a `Handle<T>` back to the `T` asset type.
```

Comment:
These docs are not super clear, especially to readers without a solid grasp of the reflection patterns.